custom_components/elasticsearch/encoder.py

```diff
@@ -0,0 +1,51 @@
+"""Custom JSON encoder for Elasticsearch."""
+
+import json
+from typing import Any
+
+from elasticsearch8.serializer import JSONSerializer
+
+
+def convert_set_to_list(data: Any) -> Any:
+    """Convert set to list."""
+
+    if isinstance(data, set):
+        output = [convert_set_to_list(item) for item in data]
+        output.sort()
+        return output
+
+    if isinstance(data, dict):
+        return {key: convert_set_to_list(value) for key, value in data.items()}
+
+    if isinstance(data, list):
+        return [convert_set_to_list(item) for item in data]
+
+    if isinstance(data, tuple):
+        return tuple(convert_set_to_list(item) for item in data)
+
+    return data
+
+
+class Serializer(JSONSerializer):
+    """JSONSerializer which serializes sets to lists."""
+
+    def json_dumps(self, data: Any) -> bytes:
+        """Serialize data to JSON."""
+
+        return json.dumps(
+            data, default=self.default, ensure_ascii=False, separators=(",", ":"), cls=Encoder
+        ).encode("utf-8", "surrogatepass")
```

Comment:
This is a copy of the function from the base class. The only difference is that I pass in `cls=Encoder`, so that we have control over set conversion within nested data structures.